### PR TITLE
Canonical orderbook API

### PR DESCRIPTION
**Pull request by p0mvn**
https://linear.app/osmosis/issue/DATA-273/expose-endpoint-for-canonical-orderbook-pools

## Background

This PR implements the new canonical orderbook API.

The pools use case provides functionality for identifying the canonical orderbook pools among all orderbook-type pools.

The canonical orderbook is defined as the orderbook with the highest liquidity for a given token pair.

During the ingestion process, as the system stores pools, it processes the orderbook pools to determine the canonical orderbook for each token pair.

The system then exposes an API that allows querying the canonical orderbook for a specific token pair as well as retrieving all canonical orderbooks for all token pairs.

## Testing

- Added unit tests
- Added integration test
- Tested with swagger against mainnet state

### Test Against Mainnet State

Get all
```bash
curl http://localhost:9092/pools/canonical-orderbooks

[
  {
    "base": "factory/osmo1z0qrq605sjgcqpylfl4aa6s90x738j7m58wyatt0tdzflg2ha26q67k743/wbtc",
    "quote": "ibc/498A0751C798A0D9A389AA3691123DADA57DAA4FE165D5C75894505B876BA6E4",
    "pool_id": 1904
  }
]
```

Get specific
```bash
curl http://localhost:9092/pools/canonical-orderbook?base=factory%2Fosmo1z0qrq605sjgcqpylfl4aa6s90x738j7m58wyatt0tdzflg2ha26q67k743%2Fwbtc&quote=ibc%2F498A0751C798A0D9A389AA3691123DADA57DAA4FE165D5C75894505B876BA6E4
1904
```

## Summary by CodeRabbit

- **New Features**
	- Added new endpoints for retrieving canonical orderbook pool IDs and entries.

- **Documentation**
	- Updated documentation with references and descriptions for new endpoints.
	- Added a reference link in `introduction.md` to direct users to more detailed information in `pools.md`.

- **Bug Fixes**
	- Corrected endpoint and method names for canonical orderbook pool ID retrieval.

- **Refactor**
	- Removed the `is_unlisted` property from the `Token` object to streamline its structure.
   
- **Tests**
	- Added new test functions for orderbook processing and pool storage functionality.
	- Introduced new methods in `poolsUseCase` struct for testing orderbook entries.

### CODE REVIEW

**Comment by p0mvn** [comment]
@coderabbitai review

File: pools/usecase/export_test.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+// WARNING: this method is only meant for setting up tests. Do not move out of export_test.go
+func (p *poolsUseCase) StoreValidOrdeBookEntry(baseDenom, quoteDenom string, poolID uint64, poolLiquidityCapitalization osmomath.Int) {
+	p.canonicalOrderBookForBaseQuoteDenom.Store(formatBaseQuoteDenom(baseDenom, quoteDenom), orderBookEntry{
+		PoolID:       poolID,
+		LiquidityCap: poolLiquidityCapitalization,
+	})
+}
+
+// WARNING: this method is only meant for setting up tests. Do not move out of export_test.go
+func (p *poolsUseCase) StoreInvalidOrdeBookEntry(baseDenom, quoteDenom string) {
```
[the remark this answers is not in the document]
Comment (cryptomatictrader) [reply pools/usecase/export_test.go]: Good catch 👍

File: domain/mocks/pools_usecase_mock.go [hunk 2]
```diff
@@ -30,6 +30,21 @@ type PoolsUsecaseMock struct {
...
+
+// GetCanonicalOrderbookPoolID implements mvc.PoolsUsecase.
+func (pm *PoolsUsecaseMock) GetCanonicalOrderbookPoolID(baseDenom string, quoteDenom string) (uint64, error) {
+	panic("unimplemented")
+}
+
+// GetCanonicalOrdrbookPoolID implements mvc.PoolsUsecase.
+func (pm *PoolsUsecaseMock) GetCanonicalOrdrbookPoolID(baseDenom string, quoteDenom string) (uint64, error) {
+	panic("unimplemented")
+}
```
[the remark this answers is not in the document]
Comment (cryptomatictrader) [reply domain/mocks/pools_usecase_mock.go]: Good catch for the typo 👍

File: pools/usecase/pools_usecase_test.go [hunk 3]
```diff
@@ -40,6 +41,8 @@ var (
 
 	defaultAmt0 = routertesting.DefaultAmt0
 	defaultAmt1 = routertesting.DefaultAmt1
+
+	defaultPoolLiduidityCap = osmomath.NewInt(100)
```
[the remark this answers is not in the document]
Comment (cryptomatictrader) [reply pools/usecase/pools_usecase_test.go]: Good catch 👍

File: pools/usecase/pools_usecase_test.go [hunk 4]
```diff
@@ -236,8 +239,263 @@
...
+		poolID                      uint64
+		poolLiquidityCapitalization osmomath.Int
+
+		preStoreValidEntryCap osmomath.Int
+		preStoreInvalidEntry  bool
+
+		expectedError   bool
+		expectedUpdated bool
+
+		expctedCanonicalOrderbookPoolID uint64
```
[the remark this answers is not in the document]
Comment (cryptomatictrader) [reply pools/usecase/pools_usecase_test.go]: Good catch 👍

**Review by cryptomatictrader — APPROVED** [review]
LGTM. Just minor typos.

**Review by deividaspetraitis — APPROVED** [review]

**Pull request merged**